connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -813,6 +823,28 @@ class SparkSession private[sql] (
    * Set to false to prevent client.releaseSession on close() (testing only)
    */
   private[sql] var releaseSessionOnClose = true
+
+  private[sql] def registerObservation(planId: Long, observation: Observation): Unit = {
+    // makes this class thread-safe:
+    // only the first thread entering this block can set sparkSession
+    // all other threads will see the exception, as it is only allowed to do this once
+    observation.synchronized {
```

Comment:
We should either lock the map, or use a concurrent hash map. Locking the observation should only be needed when we set the value, and even then the observation should be responsible for that.

Reply:
Agreed. I changed it to a concurrent hash map.